ln-gateway/src/cln.rs

```diff
@@ -40,3 +61,233 @@ pub struct HtlcAccepted {
+                amount_msat: None,
+                label: None,
+                riskfactor: None,
+                maxfeepercent: Some(max_fee_percent),
+                retry_for: None,
+                maxdelay: Some(max_delay as u16),
+                exemptfee: None,
+                localofferid: None,
+                exclude: None,
+                maxfee: None,
+                description: None,
+            }))
+            .await;
+
+        match pay_result {
+            Ok(cln_rpc::Response::Pay(pay_success)) => {
+                debug!("Successfully paid invoice");
+                let slice: [u8; 32] = pay_success.payment_preimage.to_vec().try_into().unwrap();
+                Ok(Preimage(slice))
+            }
+            Ok(_) => unreachable!("unexpected response from C-lightning"),
+            Err(cln_rpc::RpcError { code, message }) => {
+                if let Some(code) = code {
+                    debug!(%code, %message, "c-lightning pay returned error");
+                } else {
+                    debug!(%message, "c-lightning pay returned error");
+                }
+                Err(LightningError(code))
+            }
+        }
+    }
+}
+
+type PluginState = GatewayMessageSender;
+
+/// Handle core-lightning "htlc_accepted" events by attempting to buy this preimage from the federation
+/// and completing the payment
+async fn htlc_accepted_hook(
+    plugin: Plugin<PluginState>,
+    value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let htlc_accepted: HtlcAccepted = serde_json::from_value(value)?;
+    let preimage = plugin.state().send(htlc_accepted).await?;
+    let pk = preimage.to_public_key()?;
+    Ok(serde_json::json!({
+      "result": "resolve",
+      "payment_key": pk.to_string(),
+    }))
+}
+
+async fn balance_rpc(
+    plugin: Plugin<PluginState>,
+    _: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let amount = plugin.state().send(BalancePayload {}).await?;
+    Ok(json!({ "balance_msat": amount.milli_sat }))
+}
+
+async fn address(
+    plugin: Plugin<PluginState>,
+    _: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let address = plugin.state().send(DepositAddressPayload {}).await?;
+    Ok(json!({ "address": address }))
+}
+
+async fn deposit_rpc(
+    plugin: Plugin<PluginState>,
+    value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let deposit: DepositPayload = serde_json::from_value(value)?;
+    let txid = plugin.state().send(deposit).await?;
+    Ok(json!({ "fedimint_txid": txid.to_string() }))
+}
+
+async fn withdraw_rpc(
+    plugin: Plugin<PluginState>,
+    value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let withdraw: WithdrawPayload = serde_json::from_value(value)?;
+    let txid = plugin.state().send(withdraw).await?;
+    Ok(json!({ "fedimint_txid": txid.to_string() }))
+}
+
+/// Build the core-lightning plugin and register it in an LnRpc
```

Comment:
Do X and Y usually suggests that X and Y could be separate functions, which considering how long this function has to be, might be a good idea.

Reply:
Resolving this one unless I find a way of splitting out "building and registering"